[rocprofiler-systems] PyTests for TheRock

kcossett-amd opened the pull request:
## Motivation

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

Pytests for TheRock (see https://github.com/ROCm/TheRock/pull/2713)

## Technical Details

<!-- Explain the changes along with any relevant GitHub links. -->

**Migration status**

| CMake Test Group | Ported |  Notes                                                                                                                                                     |
| ---------------- | ------ | --------------------------------------------------------------------------------------------------------------------------------------------------------- |
| ROCm             | Y      | |
| rocTX            | Y      |  One failure with rocpd, however same error appears on ctest... ignoring                                                                                                                                                         |
| Hip-Stream       | Y      |                                                                                                                                                           |
| Time-Window      | Y      |                                                                                                                                                           |
| OpenMP           | Y      |                                                                                                                                                           |
| Binary Tests     | Y      |                                                                                                                                                         |
| Decode Tests | Y | |
| XGMI (GPU Connect) | Y | |
| RCCL | Y | All perfetto validation checks fail (observed in CTests). Also, the runtime tests take 10 minutes to run (and fail on TheRock).
| Config | Y | |

Think of this as phase 1, other tests will be migrated later. But as of now, we start with only this as we are limited to 15 minutes on TheRock.

**Changelog**

Install folder changes:
  - pytest package will be installed to `share/rocprofiler-systems/`
  - all compiled examples are put in `share/rocprofiler-systems/examples`
  - all test relevant scripts are put in `share/rocprofiler-systems/tests`

Non Pytest related changes:

 - Added `venv` and `.venv` to `.gitignore`
 - Moved `rocprofiler_systems_get_gfx_archs()` to `MacroUtilities.cmake` and changed it to use `rocm_agent_enumerator` as opposed to `rocminfo`
 - Added `rocprofiler_systems_lookup_gfx()`, which detects whether a given `gfx` is from the `instinct`, `radeon` or `apu` family.
 - Added `ROCPROFSYS_GFX_TARGETS` as a build argument. Used to specify the offloading architectures that GPU examples should compile for. If empty, defaults to whatever your system has.
 - GPU examples now check if the given `gfx` targets (from `ROCPROFSYS_GFX_TARGETS`) are supported.
 - All examples are installed to `share/rocprofiler-systems/examples` as opposed to `bin`
 - Fix an issue where `ROCPROFSYS_INSTALL_EXAMPLES` fails for causal tests.
 - OMPVV offload tests now only compile if `amdflang` version is `>= 20`
 - General testing scripts are installed to `share/rocprofiler-systems/tests`
 - Remove any mention of `_TRACE_WINDOW_SKIP` (unused variable)
 - Fix a problem with `rocprofsys-launch-compiler` script where it was imporperly handling `ccache`

Pytest related changes:
 - Added pytest code for some of the existing ctests (see list above).
     - For these ported tests, the behavior between ctest and pytest is intended to be identical.
 - See `README.md` in `tests/pytest/`

For now, pytests are isolated and do not interact with cmake/ctests in any way. This will be changed in the future.
No logic (as of now) is provided for cmake to detect the pytests.

**Note**
To build in such a way for examples and tests to work, use:
```
cmake -B rocprof-sys-build -S . \
    -D CMAKE_INSTALL_PREFIX=/opt/rocprofiler-systems \
    -D ROCPROFSYS_USE_PYTHON=ON -D ROCPROFSYS_BUILD_DYNINST=ON \
    -D ROCPROFSYS_BUILD_TBB=ON -D ROCPROFSYS_BUILD_BOOST=ON \
    -D ROCPROFSYS_BUILD_ELFUTILS=ON -D ROCPROFSYS_BUILD_LIBIBERTY=ON \
    -D ROCPROFSYS_BUILD_TESTING=ON -D ROCPROFSYS_INSTALL_TESTING=ON \
    -D ROCPROFSYS_INSTALL_EXAMPLES=ON
```
One can also specify `-D ROCPROFSYS_GFX_TARGETS="gfx1100;gfx1101; ...etc...`

(See `README.md` in `tests/pytest` to learn how to run the tests)

## JIRA ID

<!-- If applicable, mention the JIRA ID resolved by this PR (Example: Resolves SWDEV-12345). -->
<!-- Do not post any JIRA links here. -->

- AIPROFSYST-26

## Test Plan

<!-- Explain any relevant testing done to verify this PR. -->

(See `README.md` in `tests/pytest` to learn how to run the tests)

Tested on `gfx1100` - Pytests pass (where ctests pass)
Tested on MI300 - Pytests pass (where ctests pass) - Note that I ran into an issue where `transpose` failed as it ran out of GPU memory. Unrelated to pytests.

**Tested on TheRock (using manylinux build) - Pytests pass (requires https://github.com/ROCm/TheRock/pull/2713)**

**Note**: Running pytests with multiple processes WILL fail

## Test Result

<!-- Briefly summarize test outcomes. -->
See previous section

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

kcossett-amd committed:
Fetch Marjan's work, include OMPVV tests and fix example install locations

kcossett-amd commented:
Thank you @marantic-amd for providing the PyTest code.

kcossett-amd commented on projects/rocprofiler-systems/tests/CMakeLists.txt:
```diff
@@ -75,3 +75,39 @@ set_tests_properties(
         TIMEOUT 30
 )
+
+if(ROCPROFSYS_INSTALL_EXAMPLES)
```
Add a new variable, call it `ROCPROFSYS_INSTALL_TESTING`, that controls whether or not the logic to run tests is installed

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/conftest.py:
```diff
@@ -0,0 +1,472 @@
+            item.add_marker(skip_mpi)
+
+        if "rocpd" in item.keywords and not rocpd_available:
+            item.add_marker(skip_rocpd)
```
Might be wise to add skip conditions based on the specific card (MI300, navi). Not required but good to have.

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/gpu.py:
```diff
@@ -0,0 +1,133 @@
+        return ["GRBM_COUNT", "SQ_WAVES", "SQ_INSTS_VALU", "TA_TA_BUSY"]
+
+
+# def is_navi_architecture(arch: str) -> bool:
```
Remove

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/validators.py:
```diff
@@ -0,0 +1,376 @@
+
+# For now, pytests are only run as a solution for theRock
+# The scripts will thus be located in:
+TEST_SCRIPT_DIR = Path("/opt/rocprofiler-systems/share/rocprofiler-systems/tests/scripts")
```
Let's not have a fixed path

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/test_openmp.py:
```diff
@@ -0,0 +1,741 @@
+        # Validate timemory output contains OpenMP offloading functions
+        pattern_validation = validate_regex_patterns(
+            content=result.stdout,
+            patterns=[r"omp_offloading"],
```
Note to me future self - perhaps we can implement stricter regex checks (such as checking that the offload kernel is present for amdflang specific compilers)

kcossett-amd reviewed (commented):
Notes for myself

kcossett-amd committed:
Complete OMPVV tests, fix variable names, add print debug and add missing functions

kcossett-amd committed:
Port Transpose

kcossett-amd committed:
Remove validate_regex_patterns, add roctx tests, improve rocpd error message, and implement pass/fail regex checks

kcossett-amd committed:
Implement time window tests and remove unused variables

kcossett-amd reviewed (commented):
As a note, there should be an easy way to parametrize all the runners to avoid code repetition. For example, if we just test all of the modes with different targets and env, I shouldn't need to write the same definition for each `baselineRunner, instrumentRunner, ...etc...`

kcossett-amd committed:
Adjust pytest output dir and python validation install dir to fix pytests ran from install

kcossett-amd committed:
Add binary tests

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/test_binaries.py:
```diff
@@ -0,0 +1,758 @@
+        result = runner.run()
+        assert result.success, f"Instrument simulate lib failed: {result.failure_reason}"
+
+    # TOCHECK
```
Verify that this test mirrors cmake implementation

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/test_binaries.py:
```diff
@@ -0,0 +1,758 @@
+        result = runner.run()
+        assert result.success, f"Avail all expand keys failed: {result.failure_reason}"
+
+    # TOCHECK everything below this point
```
Verify that all tests below this point mirror cmake implementation

kcossett-amd reviewed (commented):
Ensure pytests mirror ctests.

kcossett-amd committed:
Add hip-stream tests, improve debug print helper and add rocm version fixture

kcossett-amd committed:
Port videodecode and jpegdecode tests, and fix directory install cmd

kcossett-amd committed:
Port ROCPD tests, remove rocpd fixture and integrate tests within existing tests instead of creating a class for rocpd

kcossett-amd committed:
Add ROCPROFSYS_INSTALL_TESTS and add install cmd for pytests and its dependencies

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/CMakeLists.txt:
```diff
@@ -0,0 +1,61 @@
+        ${CMAKE_CURRENT_LIST_DIR}/test_hip_stream.py
+        ${CMAKE_CURRENT_LIST_DIR}/test_jpegdecode.py
+        ${CMAKE_CURRENT_LIST_DIR}/test_openmp.py
+        ${CMAKE_CURRENT_LIST_DIR}/test_openmp.py
```
Duplicate

kcossett-amd committed:
Fix gitignore, standardize test names, add override for trace_processor_path,  add libomp pytest fixture, and make test config print before tests run

kcossett-amd committed:
Remove duplicate path and pass amdflang compiler to ompvv

kcossett-amd committed:
Add hook to print test output and add gpu fixtures to openmp tests

kcossett-amd committed:
Rename ROCPROFSYS_INSTALL_TESTS to ROCPROFSYS_INSTALL_TESTING

kcossett-amd committed:
Refactor to use pytest-subtests

kcossett-amd committed:
Make test output on failure clearer and fix fail regex not being used

kcossett-amd committed:
Add no output flag

kcossett-amd committed:
Fix order of LD_LIBRARY_PATH appending to prevent failures on TheRock

kcossett-amd committed:
Add ROCPROFSYS_GFX_TARGETS and function to classify gpu to one of instinct, radeon or apu category.

kcossett-amd committed:
Fix failure if xml output used

kcossett-amd commented:
Requires #2381 to be merged for `rocprof-sys` to work on TheRock with programs that use HSA.

kcossett-amd committed:
Skip test if target is not found

kcossett-amd committed:
Polish

kcossett-amd committed:
Add README.md, remove ROCPROFSYS_TRACE

kcossett-amd committed:
README tweak

kcossett-amd committed:
Fix function name

marantic-amd commented on projects/rocprofiler-systems/tests/pytest/CMakeLists.txt:
Do we really want to install all test files and requirements on the target system? If you take a look on my fork, the [build_standalone.sh](https://github.com/marantic-amd/rocm-systems/blob/pytest_playground/projects/rocprofiler-systems/tests/pytest/build_standalone.sh). It is a script that could wrap all of this into a single binary, removing the dependencies from the `target system environment` which might be a more graceful approach to the deployment of tests. 

I assume you saw it, so I am just curious what tipped you towards this approach?

kcossett-amd replied:
I have no idea how I missed this. Your approach is much better than what I currently have.

kcossett-amd replied:
I know that you have a manylinux option in your build script. For TheRock, I'll just use the default `shiv` option, since it works and is the default anyways. That way we test what the user would end up using.

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/gpu.py:
```diff
@@ -0,0 +1,247 @@
+    architectures: list[str]
+    device_count: int
+    is_navi: bool
+    is_mi300: bool
```
Do we really want to be this specific `is_mi300`? Maybe we can change this to be more general, `is_instinct_gpu` or something like that..

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/runners.py:
```diff
@@ -0,0 +1,756 @@
+    pass_regex: Optional[list[str]] = None
+    fail_regex: list[str] = field(
+        default_factory=lambda: ROCPROFSYS_ABORT_FAIL_REGEX.copy()
+    )
```
Do we really want to have those pass and fail regexes coupled into `TestResult`?  I would suggest extracting them as separate component, which will take TestResult as input, and could be swapped in the future with something else.

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/test_binaries.py:
```diff
@@ -0,0 +1,951 @@
+        # Check if sleep command exists
+        sleep_cmd = shutil.which("sleep")
+        if not sleep_cmd:
+            pytest.fail("sleep command not found")
```
Is test here really failed or system is not fulfilling requirements for testing? Since it is missing sleep command.

kcossett-amd replied:
Good catch, it should be `pytest.skip`

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/runners.py:
```diff
@@ -0,0 +1,756 @@
+        self.pass_regex = pass_regex
+        self.fail_regex = (
+            fail_regex if fail_regex is not None else ROCPROFSYS_ABORT_FAIL_REGEX.copy()
+        )
```
Runner should not care about regex. It should only run command and care about env and timeout.

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/runners.py:
```diff
@@ -0,0 +1,756 @@
+        # Default is False - let test_output_dir fixture handle all cleanup
+        # after validation completes
+        if self.cleanup_on_success and run_result.success:
+            if os.environ.get("ROCPROFSYS_KEEP_TEST_OUTPUT", "0") != "1":
```
This check is alredy in `cleanup_instrumented_binaries`

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/runners.py:
```diff
@@ -0,0 +1,756 @@
+        ] + self.run_args
+
+
+class SysBinaryRunner(BaseRunner):
```
`SysBinaryRunner` runner is used only in `test_binaries.py`, do we really need it? Can we reuse runners from above? This will only introduce inconsistency in tests.

kcossett-amd replied:
We do not need it. I'll merge its functionality into `BaselineRunner` since they almost do the same thing

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/test_transpose.py:
```diff
@@ -0,0 +1,638 @@
+            pytest.skip("transpose binary not found")
+
+        result = runner.run()
+        collect_result(result)
```
Since we have result here, we can create new validator, which will take regex. Regex validation will be decoupled from runner.

mradosav-amd commented on projects/rocprofiler-systems/tests/pytest/test_transpose.py:
```diff
@@ -0,0 +1,638 @@
+                pytest.fail(f"Perfetto validation failed: {validation.message}")
+
+        # ROCpd validation
+        with subtests.test("ROCpd validation"):
```
I see part of code below copied on multiple places. Would it be the issue to include this into validator itself? In that way, we can create abstraction of validators and just pass TestResult. I think it is same thing with perfetto validaton, counter validation etc.

```py
        # ROCpd validation
        with subtests.test("ROCpd validation"):
            if not use_rocpd:
                pytest.skip("ROCpd is not enabled")
            rocpd_file = result.rocpd_file
            if rocpd_file is None:
                pytest.fail(f"ROCpd database not created")
            existing_rules = [r for r in video_decode_rules if r.exists()]
            if not existing_rules:
                pytest.skip("No validation rules found")
            validation = validate_rocpd_database(
                rocpd_file,
                rocprof_config.rocprofsys_tests_dir,
                rules_files=existing_rules,
            )
            if not validation.is_valid:
                pytest.fail(f"ROCpd validation failed: {validation.message}")
```

kcossett-amd committed:
Add review suggestions.
Add fixtures that wrap validators and runners that handle pytest and subtest logic.

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/validators.py:
```diff
@@ -0,0 +1,434 @@
+        ValidationResult with validation status
+    """
+    if not trace_path.exists():
+        return ValidationResult(False, f"Trace file not found: {trace_path}")
```
I do realize that both the `assert_perfetto` fixture and `validate_perfetto_trace` check if the perfetto file exists (similar case for some other validators). 
For now, I'll leave this check here in case a future test directly uses this instead of the assert.

kcossett-amd commented on projects/rocprofiler-systems/tests/pytest/conftest.py:
```diff
@@ -0,0 +1,1151 @@
+
+# ============================================================================
+# Test run and assertion fixtures
+# ============================================================================
```
I've wrapped the validators and the runners in a fixture that also handles pytest/pytest-subtest logic, this way actual tests can be written concisely.

I do want your opinions on whether I should just move the full validate logic to the assert functions or keep them separate as they are now.

kcossett-amd committed:
Fix instinct test failure

kcossett-amd committed:
Add xgmi tests

kcossett-amd committed:
Add RCCL tests

kcossett-amd committed:
Add config tests

kcossett-amd committed:
Add check for targets that use GPU

kcossett-amd committed:
Introduce standalone build and fix GPU target detection for certain tests

kcossett-amd committed:
Add marjan's build readme, remove _check_rocm_version_gew and add sysrun_args

kcossett-amd committed:
Change the way pytest subtests failures are recorded for pytest < 9.0.0

kcossett-amd committed:
Install requirements.txt

kcossett-amd committed:
Make internal functions internal perform minor cleanups

kcossett-amd committed:
Add xdist support

kcossett-amd committed:
Revamp how output is displayed

kcossett-amd committed:
Prevent process timeouts from crashing pytest and (finally) fix --show-output

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/config.py:
```diff
@@ -0,0 +1,545 @@
+
+
+def _find_mpiexec() -> Optional[Path]:
+    """Find MPI laucnher executable."""
```
```suggestion
    """Find MPI launcher executable."""
```

kcossett-amd committed:
Add ctest-integration flag

kcossett-amd committed:
Merge branch 'develop' into users/kcossett-amd/LaRoche

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
```diff
@@ -0,0 +1,429 @@
+
+Default output directory: `<build-dir>/rocprof-sys-pytest-output/`
+
+If auto detection of the build directory fails, specify `ROCPROFSYS_BUILD_DIR=<path to build-dir>`
```
Rather than auto-detecting, consider moving these files to the "staging" area at build time, for example, `"build-dir"/share/rocprofiler-systems/tests/pytest`. I think that this is good practice and it will help to reduce discrepencies between testing a "build" and testing an "install". 
It will also keep your source directory clean from the `__pycache__` folders.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
````diff
@@ -0,0 +1,429 @@
+```bash
+python3 -m venv .venv
+source .venv/bin/activate
+pip install -r tests/pytest/requirements.txt
````
We could also move this too - `projects/rocprofiler-systems/requirements-test.txt` to make it more visible.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/CMakeLists.txt:
```diff
@@ -0,0 +1,57 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
```
I suggest using the simplied header to reduce future maintanence. For example, the year is already out of date. :)

```suggestion
# Copyright (c) Advanced Micro Devices, Inc.
# SPDX-License-Identifier:  MIT
```

kcossett-amd replied:
The joys of working on a PR between years 😄

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/test_rccl.py:
```diff
@@ -0,0 +1,204 @@
+# MIT License
```
Let's remove / ignore the RCCL profiling tests for this PR and move it to phase 2.

kcossett-amd replied:
As of now, they can be deselected by using `-k "not TestRCCL"`.
I'm currently in the process of adding markers (essentially porting ctest labels) for the workflows. Once that is done, we'll be able to use `-m`

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
```diff
@@ -0,0 +1,429 @@
+| `ROCPROFSYS_BUILD_DIR` | Path to build directory | Auto-detected |
+| `ROCPROFSYS_INSTALL_DIR` | Path to install prefix (enables install mode) | Not set |
+| `ROCPROFSYS_SOURCE_DIR` | Path to source directory | Auto-detected |
+| `ROCPROFSYS_KEEP_TEST_OUTPUT` | Keep test output on success (`ON`/`OFF`) | `OFF` |
```
I would prefer to keep this "ON" as the default. They're handy to keep around.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
````diff
@@ -0,0 +1,429 @@
+ROCPROFSYS_INSTALL_DIR=/opt/rocprofiler-systems pytest /opt/rocprofiler-systems/share/rocprofiler-systems/tests/pytest/
+```
+
+Default output directory: `/tmp/rocprof-sys-pytest-output/`
````
You will see some "false" errors on shared environments with this as your default. Consider using `/tmp/$USER/rocprof-sys-test-output` instead.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
```diff
@@ -0,0 +1,429 @@
+| Main test fails | pytest captures | pytest captures | pytest captures |
+| Subtest fails | ❌ | ✅ | ✅ |
+
+> **Perfetto GLIBC Issue:** If Perfetto validation fails due to GLIBC version mismatch, set `ROCPROFSYS_TRACE_PROCESSOR_SHELL` to a compatible binary.
```
Some additional info - This may happen on distros like RHEL-8.x and SUSE-15.5.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/README.md:
```diff
@@ -0,0 +1,429 @@
+|------|-------------|
+| `--show-output` | Show runner output when tests **pass** |
+| `--show-output-on-subtest-fail` | Show runner output only when **subtests** fail |
+| `--no-output` | Suppress all output (only show pass/fail) |
```
It would also be helpful to have results output to a log file, like `--log-file=/path/to/log-file/pytest-output.txt`, with the default being `/path/to/test-folder/pytest-output.txt`. This will be easier for use to parse data from failure, get artifacts from the CI, and QA to file tickets.

dgaliffiAMD reviewed (commented):
Thank you for your change @kcossett-amd. Very helpful README file.
I'm still reviewing, but here are a couple of high-level thoughts.

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/test_binaries.py:
```diff
@@ -0,0 +1,773 @@
+        empty_cfg = config_dir / "empty.cfg"
+        empty_cfg.write_text("#\n# empty config file\n#\n")
+
+        tmpdir = test_output_dir / "tmpdir"
```
--tmp-dir should be an absolute path

```suggestion
        tmpdir = tmpdir.resolve()
```

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/rocprofsys/config.py:
```diff
@@ -0,0 +1,545 @@
+        Returns:
+            LD_LIBRARY_PATH string with rocprofiler-systems libraries
+        """
+        paths = [str(self.rocprofsys_lib_dir)]
```
Provide absolute path for `LD_LIBRARY_PATH`. I was seeing tests fails because of this - you might have already had it in your "env", so didn't see it.
You might as well save all of the "_dir_" members this way right when they're initialized. 

```suggestion
        paths = [str(self.rocprofsys_lib_dir.resolve())]
```

kcossett-amd replied:
For the config `dirs`, I'll update only `build_dir` and `install_dir` since every other `dir` inherits that path.

kcossett-amd committed:
QOL + Aesthetic changes
 - Change copyright notices to short form

CMAKE Changes
 - Move requirements.txt to CMAKE_SOURCE_DIR
 - Move requirements.txt install location to share/rocprofiler-systems
 - During build, move tests/pytests/ to build folder
 - Use RCCL's GPU_TARGETS to specify GPUs to compile for. This reduces link time.
 - Add custom targets to move pytests to build folder
 - Remove rocprof-sys-tests-config from being installed

Pytest Changes
 - Fix environments so that they use ROCPROFSYS_TRACE
 - Remove rocprof-sys-tests-config from being cleaned up
 - Remove rocprofsys_root_dir (unused due to moving pytest to build folder)
 - Remove unused root_dir and build_dir fixtures
 - Remove cleanup_temp_files fixture (its logic is duplicate)
 - Append $USER to tmp folder location

Script Changes
 - Added xdist to build_standalone.sh

kcossett-amd committed:
QOL + Aesthetic changes
 - Update README.md
 - Add more information on GLIBC issue

Pytest Changes
 - Add --output-log flag and log file (pytest-output.txt)
 - Add --monochrome flag to append ROCPROFSYS_MONOCHROME to runner for cleaner log output
 - Add a whole bunch of labels
 - Add pytest_sessionstart section for handle terminal output redirection (log)
 - When running the package, default to install config (detect if running from tmp folder)

kcossett-amd committed:
Fix markers

kcossett-amd commented:
Testing some changes to get pytest to work on workflows

kcossett-amd committed:
Apply review suggestions, fix --no-output flag and add rccl dummy gpu to link against

kcossett-amd commented:
Note: Currently broken on python 3.7 and 3.8. I'm in the process of fixing. Also do note that python 3.6 is not supported.

kcossett-amd committed:
Python 3.7/3.8 compatibility and add xdist note in README

kcossett-amd committed:
Allow usage of roc-obj-ls in older ROCm versions to support target gfx detection

kcossett-amd committed:
Organize conftest.
Add --print-env
modify --monochrome, modify README.

kcossett-amd committed:
Merge branch 'develop' into users/kcossett-amd/LaRoche

kcossett-amd committed:
Ensure launch-compiler handles ccache properly

dgaliffiAMD commented on projects/rocprofiler-systems/tests/pytest/conftest.py:
```diff
@@ -0,0 +1,1397 @@
+        "=" * 70,
+        f"  ROCm version:   {rocm_version}",
+        f"  ROCm path:      {rocprof_config.rocm_path}",
+        f"  Is installed:   {rocprof_config.is_installed}",
```
Added some additional logs here, which we may find useful in the header.

```suggestion
        f"  ROCm version:       {rocm_version}",
        f"  ROCm path:          {rocprof_config.rocm_path}",
        f"  Is installed:       {rocprof_config.is_installed}",
        f"  Log file:           {log_file_path if log_file_path else '(not configured)'}",
        f"  Validate ROCPD:     {check_use_rocpd()}",
        f"  Validate Perfetto:  {check_use_perfetto()}",
```

where `log_file_path = getattr(config, "_output_log_path", None)`

dgaliffiAMD reviewed (commented):
Do you think these additions to the header would be useful?

kcossett-amd committed:
Add --output-dir flag
Allow --output-log to be none
Add more information to test header

kcossett-amd committed:
Add subtest_disable marker (and use it to disabled roctx_sampling_rocpd_validation).
Add --allow-disabled flag to enable these disabled tests to run.

dgaliffiAMD commented:
I think we can start to split up the PR into smaller chunks to make things more manageable. 
For example, you can use the below branch to isolate the install-path change: 
https://github.com/ROCm/rocm-systems/tree/users/dgaliffi/update-example-install-path

kcossett-amd committed:
Add --ci-mode

kcossett-amd commented:
Will close PR. To be broken up into multiple pieces.

kcossett-amd committed:
Merge branch 'develop' into users/kcossett-amd/LaRoche

kcossett-amd commented:
Closing as Part 3 covers the remainder of what is present here.
